graphql_compiler/compiler/schema_gen/schema_graph.py

```diff
@@ -0,0 +1,630 @@
+            # must not have properties of type "Link".
+            _validate_non_edges_do_not_have_edge_like_properties(class_name, properties)
+
+        _validate_property_names(class_name, properties)
+
+        self._class_name = class_name
+        self._kind = kind
+        self._abstract = abstract
+        self._properties = frozendict(properties)
+        self._class_fields = frozendict(class_fields)
+
+        # In the schema graph, both vertices and edges are represented with vertices.
+        # These dicts have the name of the adjacent schema vertex in the appropriate direction.
+        #
+        # For vertex classes:
+        #   in  = the edge is attached with its head / arrow side
+        #   out = the edge is attached with its tail side
+        #
+        # For edge classes:
+        #   in  = the tail side of the edge
+        #   out = the head / arrow side of the edge
+        #
+        # For non-graph classes, these properties are always empty sets.
+        self.in_connections = set()
+        self.out_connections = set()
+
+    @property
+    def abstract(self):
+        """Return True if the represented type is abstract, and False otherwise."""
+        return self._abstract
+
+    @property
+    def class_name(self):
+        """Return the name of the type that the schema element represents."""
+        return self._class_name
+
+    @property
+    def properties(self):
+        """Return a dict of property name to property descriptor for this schema element."""
+        return self._properties
+
+    @property
+    def class_fields(self):
+        """Return a dict containing all class fields defined on the schema element."""
+        return self._class_fields
+
+    @property
+    def is_vertex(self):
+        """Return True if the schema element represents a vertex type, and False otherwise."""
+        return self._kind == SchemaElement.ELEMENT_KIND_VERTEX
+
+    @property
+    def is_edge(self):
+        """Return True if the schema element represents an edge type, and False otherwise."""
+        return self._kind == SchemaElement.ELEMENT_KIND_EDGE
+
+    @property
+    def is_non_graph(self):
+        """Return True if the schema element represents a non-graph type, and False otherwise."""
+        return self._kind == SchemaElement.ELEMENT_KIND_NON_GRAPH
+
+    def freeze(self):
+        """Make the SchemaElement's connections immutable."""
+        self.in_connections = frozenset(self.in_connections)
+        self.out_connections = frozenset(self.out_connections)
+
+    def __str__(self):
+        """Stringify the SchemaElement."""
+        return (
+            'SchemaElement({}, {}, abstract={}, properties={}, in_conn={}, out_conn={})'
+            .format(self._class_name, self._kind, self._abstract, self._properties,
+                    self.in_connections, self.out_connections)
+        )
+
+    __repr__ = __str__
+
+
+class SchemaGraph(object):
```

Comment:
The compiler has multiple backends, and could have multiple schema generation methods for all of them. It might make sense to either create a directory for orientdb, or prefix some classnames with `OrientDB`. Also maybe take this out of the `compiler` directory to keep the hierarchy shallow.